src/datadog/grok/filters/keyvalue.rs

```diff
@@ -125,27 +179,40 @@ pub fn apply_filter(value: &Value, filter: &KeyValueFilter) -> Result<Value, Gro
     match value {
         Value::Bytes(bytes) => {
             let mut result = Value::Object(BTreeMap::default());
-            parse(
-                String::from_utf8_lossy(bytes).as_ref(),
-                &filter.key_value_delimiter,
-                &filter
-                    .field_delimiters
-                    .iter()
-                    .map(|s| s.as_str())
-                    .collect::<Vec<&str>>(),
-                &filter.quotes,
-                &filter.value_re,
-            )
-            .unwrap_or_default()
-            .into_iter()
-            .for_each(|(k, v)| {
-                if !(v.is_null()
-                    || matches!(&v, Value::Bytes(b) if b.is_empty())
-                    || k.trim().is_empty())
-                {
-                    result.insert(crate::path!(&k), v);
-                }
-            });
+            let value = String::from_utf8_lossy(bytes);
+            filter
+                .re_pattern
+                .captures_iter(value.as_ref())
+                .for_each(|c| {
+                    let key = parse_key(extract_capture(&c, 1), filter.quotes.as_slice());
+                    if key.contains(' ') {
+                        return;
+                    }
+                    let value = extract_capture(&c, 2);
+                    // trim trailing comma for value
+                    let value = value.trim_end_matches(|c| c == ',');
+
+                    if let Ok((_, value)) = parse_value(value, filter.quotes.as_slice()) {
+                        if !(value.is_null()
+                            || matches!(&value, Value::Bytes(b) if b.is_empty())
+                            || key.is_empty())
+                        {
+                            let path = crate::path!(key);
+                            match result.get(path).cloned() {
+                                Some(Value::Array(mut values)) => {
+                                    values.push(value);
+                                    result.insert(path, values);
+                                }
+                                Some(prev) => {
+                                    result.insert(path, Value::Array(vec![prev, value]));
+                                }
+                                None => {
+                                    result.insert(path, value);
+                                }
+                            };
+                        }
+                    }
```

Comment:
I think this inner block is complex enough to warrant its own function. Is that reasonable? i.e. something like `.for_each(|c| filter.try_parse_value(&mut result, c))` or whatever other parameters it needs.

Reply:
Good idea, updated https://github.com/vectordotdev/vrl/pull/1015/commits/401c44eca171b03f1c5ef47379cb622b46f0129e